oci/archive/oci_dest.go

```diff
@@ -0,0 +1,188 @@
+package archive
+
+import (
+	"io"
+	"os"
+
+	"github.com/containers/image/types"
+	"github.com/containers/storage/pkg/archive"
+	"github.com/pkg/errors"
+)
+
+type ociArchiveImageDestination struct {
+	ref          ociArchiveReference
+	unpackedDest types.ImageDestination
+	tempDir      string
+}
+
+// newImageDestination returns an ImageDestination for writing to an existing directory.
+func newImageDestination(ref ociArchiveReference, ctx *types.SystemContext) (types.ImageDestination, error) {
+	tempDirOciRef, err := createOciRef(ref.image)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error creating oci reference")
+	}
+	unpackedDest, err := tempDirOciRef.ociRef.NewImageDestination(ctx)
+	if err != nil {
+		if err := deleteTempDir(); err != nil {
+			return nil, errors.Wrapf(err, "error deleting temp directory", tempDirRef.tempDirectory)
+		}
+		return nil, err
+	}
+	return &ociArchiveImageDestination{ref: ref,
+		unpackedDest: unpackedDest,
+		tempDir:      tempDirOciRef.tempDirectory}, nil
+}
+
+// Reference returns the reference used to set up this destination.
+func (d *ociArchiveImageDestination) Reference() types.ImageReference {
+	return d.ref
+}
+
+// Close removes resources associated with an initialized ImageDestination, if any
+// Close deleted the temp directory of the oci-archive image
+func (d *ociArchiveImageDestination) Close() error {
+	if err := d.unpackedDest.Close(); err != nil {
+		if err := deleteTempDir(); err != nil {
+			return errors.Wrapf(err, "error deleting temp directory", tempDirRef.tempDirectory)
+		}
+		return errors.Wrapf(err, "error closing image destination")
+	}
+	if err := deleteTempDir(); err != nil {
+		return errors.Wrapf(err, "error deleting temp directory", tempDirRef.tempDirectory)
+	}
+	return nil
+}
+
+func (d *ociArchiveImageDestination) SupportedManifestMIMETypes() []string {
+	return d.unpackedDest.SupportedManifestMIMETypes()
+}
+
+// SupportsSignatures returns an error (to be displayed to the user) if the destination certainly can't store signatures
+func (d *ociArchiveImageDestination) SupportsSignatures() error {
+	err := d.unpackedDest.SupportsSignatures()
+	if err != nil {
+		if err := deleteTempDir(); err != nil {
```

Comment:
`SupportsSignatures` returning an error is not a critical failure, no need to clean stuff up just because `SupportsSignatures` returns an error. Besides, `Close` will do the cleanup anyway (the caller is expected to ensure that `Close` is eventually called).